Code translated from MATLAB and ABOUTTHECODE markdown added.

jsmatias opened the pull request:
@cacarrara, sorry the late reply. 
Here is the code moved from: https://github.com/sosbrumadinho/brumadinho_location
o/

jsmatias committed:
Code translated from MATLAB and ABOUTTHECODE markdown added.

jsmatias committed:
Vectorial field initial stage translated from Matlab code

volpatto commented:
@jsmatias could you provide some explanation about what this code does? I took a look but don't get it. The code uses external results such as CFD simulation computed velocity fields?

jsmatias committed:
Improved code documentation and add matlab files

Second improvement of the MATLAB code was added to the matlab_folder.
The translation to Python evolved.

jsmatias committed:
Typo in the readme file

Typo corrected

jsmatias committed:
About to finish matlab code translation

Only the last part of the main.m file still has to be converted.

jsmatias committed:
First MVP

The code has to be tested and validated. Using a initial position it calculates force, acceleration, velocity and position arrays of the possible path of an object being drifted by the fluid.

jsmatias commented:
> @jsmatias could you provide some explanation about what this code does? I took a look but don't get it. The code uses external results such as CFD simulation computed velocity fields?

The original code was written in MATLAB and then translated to Python. All the MATLAB code can be found in the matlab_code folder. 

The main.py file gets the geolocation coordinates and converts to x and y positions, which are used as an initial position. Data in the data folder are used to identify the the region affected by the accident. This region is passed to the FLUID class which calculates the velocity field and force field. As follows, using an average mass equals 80kg (average person), the acceleration field is obtained.  
After two integrations of the acceleration field, the code determines the positions of the body. The final position is then determined using the last value in the position array. All this info and more details are available in the ABOUTTHECODE.md file. 

@volpatto Could you have a look again on the code? 
Let me know if you have any question. 
If you run  the main file you should have a plot and a final position as shown here:

This is a representation of the mud flowing. 
![image](https://user-images.githubusercontent.com/29777926/52445871-42bcd780-2b24-11e9-8963-e30fd4e675ba.png)

Thank you guys for the reviews!
cc.: @volpatto @pictos

pictos commented:
@volpatto, can you review this PR for us?

volpatto commented:
> @volpatto, can you review this PR for us?

@pictos sure, I will take it after work ASAP, today or tomorrow. @jsmatias I will check your code, now it seems to be more complete.

volpatto commented on utils/fluid.py:
```diff
@@ -0,0 +1,39 @@
+import pandas as pd
+import utm
+
+class FLUID():
```
Perhaps you could employ here [attrs](https://github.com/python-attrs/attrs) lib to implement classes. Another point is the notation, very uncommon for Python's Classes. I think you could apply a more rigorous format. There is [yapf](https://github.com/google/yapf) as an option to format the code.

volpatto commented on utils/fluid.py:
```diff
@@ -0,0 +1,39 @@
+  _velocityField = pd.DataFrame(columns = ['vLon', 'vLat'])
+  _forceField = pd.DataFrame(columns = ['fLon', 'fLat'])
+
+  def MaxVelocity(self, time):
```
If you don't mind, please try to use some Python classical formats (like pep8) for methods as well

jsmatias replied:
I run autopep8 to the code. I guess it's much better now.

volpatto commented on utils/geomap.py:
```diff
@@ -0,0 +1,98 @@
+      for col in range(1,cols-1):
+        lat = data.loc[(row, 'lat'), col]
+        lon = data.loc[(row, 'lon'), col]
+
```
I think that maybe all of those indices can be employed in a vectorized fashion, just like [here](https://nbviewer.jupyter.org/github/opesci/devito/blob/master/examples/cfd/01_convection.ipynb). Compare the second equation with the cell 3 in the link. I would recommend that convert the `pandas.DataFrame` to `numpy.ndarray` and use vectorization. In general, the gain in performance can reach 100~300 times faster. The MDT gradient uses a kind of stencil, very much like Finite Difference Method. So it's possible to vectorize it.

volpatto commented on utils/geomap.py:
```diff
@@ -0,0 +1,98 @@
+        dh_dp = np.sqrt( dz_dx**2 + dz_dy**2)
+
+        mod = dh_dp
+        ang = math.atan(dz_dx/dz_dy) + math.pi
```
If you vectorize your code, you will need to replace all `math` function to NumPy's Universal Functions, suitable for `ndarray`s as input

volpatto commented on test/test.py:
```diff
@@ -0,0 +1,62 @@
+# Loads borders of the region 
+borders = pd.read_csv('./data/borders.csv')
+# converts borders to utm and append it to original borders
+borders[ ['x0', 'y0', 'utmZoneNumber', 'utmZoneLetter'] ] = borders.apply( 
```
Minor: I think this is an uncommon format. Too much empty spaces.

volpatto commented on test/utils/vector_field.py:
```diff
@@ -0,0 +1,89 @@
+
+    utmXYProj = np.zeros((rows, cols, 2))
+
+    for row in range(1,rows-1):
```
As I said in another file (this also indicates duplicated code, an undesired code feature), I think you can pass it all to a `numpy.ndarray` and perform vetorized operations instead loops. The gain in time is outstanding.

jsmatias replied:
Removed test folder as it was duplicated.

volpatto commented on test/map.py:
```diff
@@ -0,0 +1,88 @@
+      self._data
+    )
+
+  def MdtGradient(self, data, step):
```
Duplicated code. Perhaps the MDT deserves a module for itself.

jsmatias replied:
Thank you for the review! 
All the files inside the test folder is duplicated. I had a problem with my laptop, so I'm using another machine which I can't install git. I tried to delete that folder, however I couldn't find a way to do it here in the Github page. Is it possible?
Anyway, I might get my new laptop this week, then I'll fix this problem. 

Thanks again!

volpatto commented on test/geomap.py:
```diff
@@ -0,0 +1,119 @@
+  #     self._data
+  #   )
+
+  def MdtGradient(self, data, step):
```
Ditto. I think there is too much similar codes, it's not clear to me the division you're doing in your code. Looks like there is a directory for tests (not in the sense of software development), but I think you shouldn't include it in the codebase. Perhaps you can perform it only locally.

jsmatias replied:
Removed duplicated code.

volpatto reviewed (changes requested):
I think the code has an initial form and this is good. Some work need to be done, but congratulations for what is done so far. I will take a look in the localization method in the next few days, I think it can be improved.

jsmatias committed:
Clean up and pep8 lint the code

volpatto commented:
@jsmatias thanks for your changes. I will take a look again ASAP. I think there is not too much to do for improving the code, but it's always worth to check if it's possible.
